### PR TITLE
Feature/advanced message filters

### PR DESCRIPTION
**Pull request by roman-khomenko-dev**
Write advanced filtering functionality and cover by tests

### CODE REVIEW

File: lib/chatter_web/live/chat_live/index.ex [hunk 1]
```diff
@@ -111,4 +137,72 @@ defmodule ChatterWeb.ChatLive.Index do
...
+    case option do
+      :with_likes_who_liked -> filter_with_likes_who_like()
+      :without_likes_who_never_liked -> filter_without_likes_who_never_liked()
+      :with_major_likes -> filter_with_major_likes()
+    end
+  end
+
+  defp filter_with_likes_who_like do
+    Enum.filter(MessageAgent.get(), fn message ->
+      if Enum.count(message.likes) > 0 && message.author in MessageAgent.get_all_likes(),
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:151]: `N+1` problem here
it gets all likes authors for each message

can be improved

File: lib/chatter_web/live/chat_live/index.ex [hunk 2]
```diff
@@ -111,4 +137,72 @@ defmodule ChatterWeb.ChatLive.Index do
...
+
+      if acc.percent < 80.0, do: {:cont, acc}, else: {:halt, acc}
+    end)
+  end
+
+  defp messages_with_likes_percent({messages, likes_summary}) do
+    messages
+    |> Enum.reduce([], fn message, acc ->
+      acc ++
+        [
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:190]: reduce( ...
[%{} | acc]
...)
|> Enum.reverse()

File: lib/chatter_web/live/chat_live/advanced_filter_component.ex [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  use ChatterWeb, :live_component
+
+  def render(assigns) do
+    ~H"""
+    <div class="dropdown d-grid gap-2">
+      <button class="btn btn-sm btn-dark dropdown-toggle" type="button" data-bs-toggle="dropdown" data-bs-auto-close="false" aria-expanded="false">
+        Advanced filters
+      </button>
+      <ul class="dropdown-menu">
+        <li><a class={"dropdown-item #{if @filter_option == :with_likes_who_liked do "active" end}"} phx-click="filter_option" phx-value-option="with_likes_who_liked">
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/advanced_filter_component.ex]: added live_helpers with `classes` and used it

File: lib/chatter_web/live/chat_live/index.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex]: right, getting all likes is moved to the top of function

File: lib/chatter_web/live/chat_live/index.ex (hunk 2 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex]: knows now about performance issue with appending to list with list ++ value, fixed

File: lib/chatter_web/live/chat_live/index.ex [hunk 4]
```diff
@@ -111,4 +137,72 @@ defmodule ChatterWeb.ChatLive.Index do
...
+              |> calculate_likes_percent(likes_summary)
+          }
+        ]
+    end)
+  end
+
+  defp calculate_likes_percent(message_likes, likes_summary) do
+    message_likes
+    |> Decimal.div(likes_summary)
+    |> Decimal.to_float()
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex:205]: i see, message_likes / likes_summary * 100 works same

**Pull request merged**